File: Sources/ExcalidrawMath/Vector.swift
import Foundation

/// A 2D vector (`packages/math/src/vector.ts`).
public struct Vector: Equatable, Hashable, Sendable {
    public var u: Double
    public var v: Double

    public init(_ u: Double, _ v: Double) {
        self.u = u
        self.v = v
    }

    /// Vector from `origin` to `point` (`vectorFromPoint`).
    public init(from point: Point, origin: Point = .zero) {
        self.init(point.x - origin.x, point.y - origin.y)
    }

    public static let zero = Vector(0, 0)

    public static func + (lhs: Vector, rhs: Vector) -> Vector {
        Vector(lhs.u + rhs.u, lhs.v + rhs.v)
    }

    public static func - (lhs: Vector, rhs: Vector) -> Vector {
        Vector(lhs.u - rhs.u, lhs.v - rhs.v)
    }

    public func scaled(by scalar: Double) -> Vector {
        Vector(u * scalar, v * scalar)
    }

    /// 2D cross product (the scalar z-component) (`vectorCross`).
    public func cross(_ other: Vector) -> Double {
        u * other.v - other.u * v
    }

    /// Dot product (`vectorDot`).
    public func dot(_ other: Vector) -> Double {
        u * other.u + v * other.v
    }

    public var magnitudeSquared: Double { u * u + v * v }

    public var magnitude: Double { magnitudeSquared.squareRoot() }

    /// Unit vector, or the zero vector when the magnitude is zero (`vectorNormalize`).
    public func normalized() -> Vector {
        let m = magnitude
        return m == 0 ? .zero : Vector(u / m, v / m)
    }

    /// Right-hand normal (`vectorNormal`).
    public func normal() -> Vector {
        Vector(v, -u)
    }

    /// The point this vector points at from `offset` (`pointFromVector`).
    public func point(offset: Point = .zero) -> Point {
        Point(offset.x + u, offset.y + v)
    }
}

File: web/packages/math/src/vector.ts
import { Point } from "./point.js";

/** A 2D vector. (parity: Sources/ExcalidrawMath/Vector.swift) */
export class Vector {
  constructor(
    public u: number,
    public v: number,
  ) {}

  static readonly zero = new Vector(0, 0);

  /** Vector from `origin` to `point` (`vectorFromPoint`). */
  static fromPoint(point: Point, origin: Point = Point.zero): Vector {
    return new Vector(point.x - origin.x, point.y - origin.y);
  }

  add(other: Vector): Vector {
    return new Vector(this.u + other.u, this.v + other.v);
  }

  sub(other: Vector): Vector {
    return new Vector(this.u - other.u, this.v - other.v);
  }

  scaled(scalar: number): Vector {
    return new Vector(this.u * scalar, this.v * scalar);
  }

  /** 2D cross product (scalar z-component) (`vectorCross`). */
  cross(other: Vector): number {
    return this.u * other.v - other.u * this.v;
  }

  /** Dot product (`vectorDot`). */
  dot(other: Vector): number {
    return this.u * other.u + this.v * other.v;
  }

  get magnitudeSquared(): number {
    return this.u * this.u + this.v * this.v;
  }

  get magnitude(): number {
    return Math.sqrt(this.magnitudeSquared);
  }

  /** Unit vector, or zero when magnitude is zero (`vectorNormalize`). */
  normalized(): Vector {
    const m = this.magnitude;
    return m === 0 ? Vector.zero : new Vector(this.u / m, this.v / m);
  }

  /** Right-hand normal (`vectorNormal`). */
  normal(): Vector {
    return new Vector(this.v, -this.u);
  }

  /** The point this vector points at from `offset` (`pointFromVector`). */
  point(offset: Point = Point.zero): Point {
    return new Point(offset.x + this.u, offset.y + this.v);
  }
}
